code/site/components/com_pages/dispatcher/router/resolver/page.php

```diff
@@ -0,0 +1,98 @@
+<?php
+/**
+ * Joomlatools Pages
+ *
+ * @copyright   Copyright (C) 2018 Johan Janssens and Timble CVBA. (http://www.timble.net)
+ * @license     GNU GPLv3 <http://www.gnu.org/licenses/gpl.html>
+ * @link        https://github.com/joomlatools/joomlatools-pages for the canonical source repository
+ */
+
+class ComPagesDispatcherRouterResolverPage extends ComPagesDispatcherRouterResolverHttp
+{
+    protected function _initialize(KObjectConfig $config)
+    {
+        $config->append(array(
+            'priority' => self::PRIORITY_NORMAL,
+            'routes'   => $this->getObject('page.registry')->getRoutes()
+        ));
+
+        parent::_initialize($config);
+    }
+
+    public function resolve(ComPagesDispatcherRouterInterface $router)
+    {
+        if($route = parent::resolve($router))
+        {
+            $response = $router->getResponse();
+            $path     = $route->getPath();
+
+            if($page = $this->getObject('page.registry')->getPage($path))
```

Comment:
@johanjanssens Same feeling here as with the setCannonicalUrl call made below. When the router resolves a request it should only return a route. I believe the resolver is trying to do too much here. This whole block should perhaps be moved outside the router package and put in the pages dispatcher.

Reply:
See comment above. You are missing the point what a router actually does. A router is not just url matcher or url generator, it has a wider responsibility.